src/hooks/useOptimisticUpdating.js

```diff
@@ -0,0 +1,54 @@
+import { useCallback, useEffect, useRef } from 'react';
+
+import { debounce } from '../utilities';
+
+/**
+ * Custom hook to be used to optimistically update a TextInput with direct manipulation
+ * before invoking the passed onChange callback.
+ *
+ * @param {Any}  value          Generic value of some state
+ * @param {Func} onChange       Function to call when a change event occurs
+ * @param {Func} preChangeHook  Function to call when a change event occurs, before setting state.
+ * @param {Func} postChangeHook Function to call after setting state after a change event
+ */
+export const useOptimisticUpdating = (value, onChange, preChangeHook, postChangeHook) => {
+  // Ref to pass to the underlying text input which enables direct manipulation of values
+  // as triggering state updates every few ms with the long press feature is not performant.
+  const ref = useRef();
+
+  // Keep an optimistic state as a ref which when updating will not trigger re-renders at the
+  // react level. Update the optimistic value with any values passed as props to ensure this
+  // component is still a controlled input.
+  const optimisticValue = useRef(value);
+  useEffect(() => {
+    optimisticValue.current = value;
+  }, [value]);
+
+  // Wrapping the onChange callback such that it is only triggered every second attempts to
+  // keep the real source of truth in sync with this optimistic state being used for more
+  // performant drawing in the interface.
+  const wrappedOnChange = useCallback(
+    debounce(newValue => {
+      onChange(newValue);
+    }, 500),
+    []
+  );
+
+  // When changing, adjust the optimistic state and directly manipulate
+  // the TextInput - this draws to the screen without causing a full re-render cycle in
+  // React. Then, attempt to update the source of truth.
+  const withOptimisticState = addend => {
+    const newValue = preChangeHook(optimisticValue.current, addend);
+    optimisticValue.current = newValue;
+    ref?.current?.setNativeProps({ text: postChangeHook(optimisticValue.current) });
+    return newValue;
+  };
+
+  const newOnChange = newValue => {
+    wrappedOnChange(withOptimisticState(newValue));
+  };
+
+  const newValue = postChangeHook(optimisticValue.current);
+
+  return [ref, newValue, newOnChange];
+};
```

Comment:
Whole thing basically returns a ref which you apply to a text input, then wraps the onChange for the textinput and directly updates the dom element and debounces the real callback.

reasonably common use case to optimistically update state and think this is a pretty OK solution. This is the main change of the PR, improving the stepper performance

Reply:
not sure if it is because of this.. the stepper is behaving very strangely now. default of 5mins. Click `+` and it goes to 6 mins. click `-` and it goes to 1 mins. Go up again and it is 3537 mins and then keeps going up in about 3000 increments

Reply:
other than that, the wrapper looks nifty. nice idea.